SDK/MobileCenter/Microsoft.Azure.Mobile.Windows.Shared/Utils/Synchronization/StatefulMutex.cs

```diff
@@ -11,16 +11,53 @@ namespace Microsoft.Azure.Mobile.Utils.Synchronization
     /// </summary>
     public class StatefulMutex : IDisposable
     {
+        private const string ShutdownExceptionMessage = "Trying to execute task after shutdown requested";
+        private const string StateExceptionMessage = "Cannot lock mutex with expired state";
+
         private readonly SemaphoreSlim _mutex = new SemaphoreSlim(1, 1);
-        private readonly StateKeeper _stateKeeper;
+        private State _state = new State();
+        private volatile bool _isShutdown;
+
+        /// <summary>
+        /// Gets the current state
+        /// </summary>
+        /// <returns>The current state</returns>
+        public State State
+        {
+            get { return _state; }
+        }
 
         /// <summary>
-        /// Creates a <see cref="StatefulMutex"/>  with an existing <see cref="StateKeeper"/> 
+        /// Advances the current state
         /// </summary>
-        /// <param name="stateKeeper">Tracks the current state for the mutex</param>
-        public StatefulMutex(StateKeeper stateKeeper)
+        public void InvalidateState()
         {
-            _stateKeeper = stateKeeper;
+            _state = _state.GetNextState();
+        }
+
+        /// <summary>
+        /// Checks if the given state is current
+        /// </summary>
+        /// <param name="state">The state to test</param>
+        /// <returns>True if the current state is current, false otherwise</returns>
+        public bool IsCurrent(State state)
+        {
+            return _state.Equals(state);
+        }
+
+        public async Task ShutdownAsync()
```

Comment:
https://msdn.microsoft.com/en-us/library/system.threading.semaphoreslim(v=vs.110).aspx 

"If multiple threads are blocked, there is no guaranteed order, such as FIFO or LIFO, that controls when threads enter the semaphore. "